### PR TITLE
WIP: Possible fix for numba bug.

### PR DESCRIPTION
**Pull request by hoytak**
This bug is extremely interesting.  It appears to be a numba bug revealed in the cfunc path of numba, which is exposed when numpy is used (as documented in https://github.com/numba/numba/issues/2904. ) This bug is exposed when the a thread other than the main thread on which the python interpreter is started is actually the thread executing the function.  When this happens, PyGILState_Ensure() goes down a different path and attempts to acquire a lock on the interpreter, but the main thread never releases that lock.  

The workaround is to swap the thread states of the interpreter between the evaluation thread and the main thread before evaluation calls to the lambda function.

### CODE REVIEW

File: src/unity/python/turicreate/test/test_lambda.py [hunk 1]
```diff
@@ -76,3 +76,20 @@ def bad_fun(x):
...
         glconnect.get_unity().parallel_eval_lambda(lambda x: good_fun(x), ls)
+
+    def test_numba(self):
+
+        from numba import jit
+        import numpy as np
+        import turicreate as tc
+
+        @jit
+        def _sum(x):
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_lambda.py]: Probably want to call this something other than `_sum` cause it doesn't sum, even `foo` I think would be fine for a unit test.

File: src/unity/python/turicreate/test/test_lambda.py [hunk 2]
```diff
@@ -76,3 +76,20 @@ def bad_fun(x):
...
+            return x[0, 0]
+
+        print("BORK")
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_lambda.py]: You should get rid of these two print statements.

**Comment by TobyRoseman** [comment]
@hoytak - thanks for working on this! The code looks good to me, just a couple minor things. Although someone more familiar with this part of the code base should really take a look.

I tried testing this fix with the Sound Classifier's preprocessing. However I'm even able to get to the point where I call `tc.sound_classifier.create(...)`. Running the code below out of master is pretty fast (less than a minute). However it takes at least a very long time (maybe never finishes) with your changes rebased on top of master.

The following code uses the [ESC dataset](https://github.com/karoldvl/ESC-50). `data` only has 2,000 rows.
```python
import turicreate as tc
from os.path import basename

# Load the audio data and meta data.
data = tc.load_audio('./ESC-50/audio/')
data['filename'] = data['path'].apply(lambda p: basename(p))
data.__materialize__()
```

**Comment by hoytak** [comment]
Yes, it appears this breaks other code.  It's not ready yet.

**Comment by ylow** [comment]
I am not entirely convinced that the diagnosis is right. I have tried a change where the main thread keeps reentering the Python interpreter thus should avoid any issues with the main thread holding on to the GIL. And that does not seem to work either.

**Comment by ylow** [comment]
I worry that with an incomplete understanding of the problem this fix may cause problems in different places.

**Comment by hoytak** [comment]
I'm abandoning this PR.  This is a won't fix -- there's no easy way I see to get around the numba bug.

**Pull request closed without merging**